### DISCUSSION TITLE
Run code which opens a PTY

### DISCUSSION THREAD

**Discussion by decentral1se** [charmbracelet/wish#307] [Q&A] [answered]
Hi 👋 I'm trying to figure out how a user can ssh into a wish bubbletea interface and then drop into a pty via some code that runs something like "docker exec" for a container. I'm using the the Go API bindings, so I can't use `wish.Command` to shell out. So I'm looking for some way to pass a function that opens a PTY like `wish.Command`.

**Comment by decentral1se** [comment]
I noticed that the code I want to run has this kind of args available:

```go
args := foo.InstanceExecArgs{
  Stdin:  os.Stdin,
  Stdout: os.Stdout,
  Stderr: os.Stderr,
}
```

However, I'm not sure if I can get access to that from the `pty, _, _ := sess.Pty()`? I see that `pty.Slave/Master` are `*os.File`? Not sure how to deal with `Stderr` 🤔

**Reply by aymanbagabas** [reply]
Hi @decentral1se. The [SSH library](https://pkg.go.dev/golang.org/x/crypto/ssh#Channel) handles writing to the appropriate channel. To use `Stderr` in your PTY session, use `sess.Stderr()`. See https://stackoverflow.com/questions/16838624/why-does-openssh-requesttty-cause-stderr-redirected-to-stdout/16839544#16839544

**Reply by decentral1se** [reply]
Thanks for this @aymanbagabas! I'm still struggling a bit... with:

```go
args := foo.InstanceExecArgs{
  Stdin:  pty.Master,
  Stdout: pty.Slave,
  Stderr: sess.Stderr(),
}
```

I do get the PTY in the client console but then I see it open as follows with these additional characters:

```
~ # ^[[6n
```

And typing any commands results in strange errors related to permissions in the container 🤔 Is there something I'm missing with threading the `Stdin/out`?

**Reply by aymanbagabas** [reply, marked as answer]
Hey @decentral1se, you should actually pass `pty.Slave` to all Stdio.

**Reply by decentral1se** [reply]
Oh yes, that's the one, tysm @aymanbagabas 🎉

I also had to wrangle that into another snippet which I'll share:

```go
if req.Interactive {
    cfd := int(pty.Slave.Fd())
    oldttystate, err := termios.MakeRaw(cfd)
    if err != nil {
        slog.Error(err.Error())
        os.Exit(1)
    }
    defer termios.Restore(cfd, oldttystate)
}
```
